### PR TITLE
fix: allow running tests outside Vite project root folder

### PR DESCRIPTION
**Pull request by IlCallo**
- Reference https://github.com/cypress-io/cypress/issues/22505 (I'm not sure it solves all the problems in that issue)

### Additional details
See explanation here: https://github.com/cypress-io/cypress/issues/22505#issuecomment-1426111689

Note that I'm not sure where `@fs` comes from, probably a Vite plugin?

### Steps to test
_You'll need Yarn 1, unluckily we're still stuck with that_

- clone https://github.com/IlCallo/quasar and checkout `upgrade-cypress-ae`
- `cd ui`
- `yarn install`
- `cd dev`
- `yarn install`
- `cd ..`
- `yarn test:component`
- notice Cypress can correctly detect tests
- click on any test, I did my tests on `QAvatar` ones
- notice the request returns 404, due to a wrong relative path composition

### How has the user experience changed?

We're now able to reference tests outside Vite project root folder in Quasar monorepo

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by ZachJW34** [comment]
@IlCallo Could you run `SNAPSHOT_UPDATE=1 yarn workspace @tooling/system-tests test vite_dev_server_fresh_spec` (after running `yarn` and making sure everything is built)? It should update our snapshot file and then you can push those up in a separate commit to get the tests to pass.

Also, we have an explicit `fs.allow` list https://github.com/cypress-io/cypress/blob/9db5251a0ee954b260abdd1ddf217834667818bc/npm/vite-dev-server/src/resolveConfig.ts#L111

so I'm surprised your app is working without modifying this. We might be able to add `fs.strict: false` here to just allow all. Any thoughts?

Would also be good to get a test demonstrating this new behavior which I can help you with.

**Comment by lmiller1990** [comment]
Copy pasting the reproduction @IlCallo supplied in Discord for some context

```
Notes about the repro:

- you'll need Yarn 1, unluckily we're still stuck with that
- clone https://github.com/IlCallo/quasar and checkout upgrade-cypress-ae
- cd ui
- yarn install
- cd dev
- yarn install
- cd ..
- yarn test:component
- notice Cypress can correctly detect tests
- click on any test, I did my tests on QAvatar ones
- notice the request returns 404, probably due to a wrong path 
```

Not sure if this fixes all issues in https://github.com/cypress-io/cypress/issues/22505, but it looks like it at least fixes Quasar, who ships us as part of their first class tooling, so I'd definitely like to see this one reviewed and merged (unless it is introduces a regression).

File: npm/vite-dev-server/client/initCypressTests.js [hunk 1]
```diff
@@ -32,13 +32,17 @@ if (supportFile) {
   })
 }
 
+// Using relative path wouldn't allow to load tests outside Vite project root folder
+// So we use the "@fs" bit to load the test file using its absolute path
+const testFileAbsolutePathRoute = `${devServerPublicPathRoute}/@fs${CypressInstance.spec.absolute}`
+
```
Comment (lmiller1990) [inline npm/vite-dev-server/client/initCypressTests.js:38]: Looks like `@fs` is a Vite thing - I couldn't find much documentation outside of https://vitejs.dev/config/server-options.html#server-fs-allow.

I'll update the snapshots - I don't know if this breaks anything, but I always thought the `src/src` duplication was kind of weird.

**Comment by lmiller1990** [comment]
I took care of updating the snapshots [ecdabcd](https://github.com/cypress-io/cypress/pull/25801/commits/ecdabcd68b0ce2b85dd4947addda3a266f7a2a3e)

@ZachJW34 do you have any thoughts on the blast radius of this change? I suspect this is an implementation that shouldn't impact anyone (other than people experiencing a related bug, for whom it'll fix).

**Comment by IlCallo** [comment]
> so I'm surprised your app is working without modifying this

I'm surprised too, but I'm not a Vite expert, so there's probably some underlying business rule I'm not aware of.
Could it be due to the fact that we provide our own vite dev server config?

**Review by ZachJW34 — APPROVED** [review]
@lmiller1990 I think the blast radius is small. Stack traces are working properly and it fixes the bug (checked out the quasar repro to verify). It was cool to see that the specs list works as expected for file outside the projectRoot (with the `../`)!

I'm good with this change!

**Review by lmiller1990 — APPROVED** [review]

**Comment by lmiller1990** [comment]
Thanks @IlCallo, this will go in the next release, that'll be in ~2 weeks. I'll merge it up by the end of this week (need to do a few housekeeping things, like a changelog entry, etc).

**Comment by IlCallo** [comment]
@ZachJW34 @lmiller1990 
I dug a bit further and found out why it works for Quasar repo tests: we tamper with the viteConfig of our project adding `..` into `server.fs.allow`

https://github.com/quasarframework/quasar/blob/8f189d3f6b54d083cb40da8bf809d38dd120faa5/ui/dev/quasar.config.js#L80-L82

The change in this PR is still relevant, but it may be a good idea to specify somewhere on the docs that customizing the `server.fs.allow` property is needed to access files outside what Cypress consider the project root folder

Unless there are security concerns, removing the `server.fs.allow` altogether could be an option too.
I guess it depends on why it has been added in the first place

**Comment by lmiller1990** [comment]
I suspect we should not tamper with `server.fs.allow`, since it's related to opt in file system access (I think we a third party library should not mess with this) but we can definitely add a note to the docs around this.

**Comment by IlCallo** [comment]
> suspect we should not tamper with server.fs.allow

Playing the guess game, I'd say you added that option because since Vite 2.7 that option is enabled by default and only allows to access files in the current workspace(s), but you have some files you need which are outside the project root folder (as we have), especially the Cypress binary stored in the global cache

Another related problem which recently came up is that you aren't re-adding the workspaces paths to the allow after overriding the option, which cause problems when using Cypress component testing to execute all tests into a monorepo with multiple packages: https://github.com/cypress-io/cypress/pull/26036

**Comment by mike-plummer** [comment]
@lmiller1990 Potentially related [issue](https://github.com/cypress-io/cypress/issues/26035) & [PR](https://github.com/cypress-io/cypress/pull/26036) opened to work around the `server.fs.allow` prop in monorepos, in case you're curious

**Pull request merged**